### ISSUE TITLE
TypeVars are lost when passed through some type-preserving operations

### ISSUE THREAD

**Issue by supersergiy** [microsoft/pyright#7874]
**Describe the bug**
TypeVars are lost when passed through some type-preserving operations.

**Code or Screenshots**
```python
from __future__ import annotations
from typing import Generic, TypeVar

T = TypeVar("T", int, float)

class Foo(Generic[T]):
    def __init__(self, x: T):
        self.x: T = x
        
    def foo(self) -> T:
        return self.x * 2
        
    def __neg__(self) -> Foo[T]:
        return Foo[T](x=-self.x)
```
[[pyright playground link](https://pyright-play.net/?code=GYJw9gtgBA%2BjwFcAuCQFM5QJYQA5hCSgEMA7UsJYpLMUgZwChRIokBPXLUgc2zwJEA4mlJoQWAMYAaKABVOaAGrEQjRnKgBeeYpUgAFACI5R2dySzgAGzDUAlOsnXi9elABiYMAZFiJkgDacgC69gBcjFDRUAAmaMCwMNxYSHAG9GjWwLIAHuHyEVExJZnZAHT58tpQucUl0fVxCVDA3hlZwPZQALQAfPKRDTHoKCCkUGXAlVAAVFAATE0lTfGJcGI86VPd-Z7ewSFDw1CjqBNeYIcGuVo9U5WOjEA)]

Same code does not produce a type error in mypy.

Changing the unary negation to `* -1` removes the type error in this case, but the problem is more general than that, eg `abs(x)` still produces a problem

**Comment by erictraut** [comment]
I agree this is a bug, although your usage of a constrained TypeVar here is suspect. You're constraining the type to `int` and `float`, but `int` is a subtype of `float` in the Python type system, so your constraints are effectively `int` and `float | int`. 

The [typing spec](https://typing.readthedocs.io/en/latest/spec/generics.html#introduction) is not clear about how constrained TypeVar should work when some of the constraints are subtypes of other constraints. This is something I hope to clarify at some point in the future. It's possible that this will become invalid, in which case your code will no longer work. 

This bug is manifest even if the constraints are non-overlapping types, so it's something that should be fixed.

It will be addressed in the next release of pyright.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.363, which I just published.

**Comment by supersergiy** [comment]
I hope they don't make it invalid to use constraints are subtypes of other constraints, otherwise they'll have to update the examples in https://peps.python.org/pep-0484/

<img width="414" alt="image" src="https://github.com/microsoft/pyright/assets/5151706/409e77a9-0aa8-4eb4-b0b7-55364ef0e934">
